### PR TITLE
76 autopipe adam

### PR DESCRIPTION
**Pull request by awtkns**
ququququq

@KlamChowder1 I got this implementation working for every node

### CODE REVIEW

File: backend/api/models/dicom.py [hunk 1]
```diff
@@ -22,6 +27,14 @@ class DicomNode(PathMixin, Base):
         UniqueConstraint('title', 'host', 'port', 'user_id', name='_node_uc'),
     )
 
+    @property
+    def is_rts(self):
+        return self.host == config._RTS_HOST and self.port == config._RTS_PORT
+
+    @staticmethod
+    def compare(a, b) -> bool:
```
Comment (asim-shrestha) [inline backend/api/models/dicom.py:35]: Could alternatively override __eq__  so we can do a == b directly

File: backend/api/models/pipeline.py [hunk 2]
```diff
@@ -16,12 +17,10 @@ class Pipeline(Base):
     ae_title = Column(String, unique=True)
     is_shared = Column(Boolean, default=False)
 
-    runs = relationship("PipelineRun", backref="pipeline",
-                        passive_deletes=True)
+    runs = relationship("PipelineRun", backref="pipeline", passive_deletes=True)
     nodes = relationship("PipelineNode", backref="pipeline")
     links = relationship("PipelineLink", backref="pipeline")
 
-    # TODO: This query can be optimized by joins
```
Comment (asim-shrestha) [inline backend/api/models/pipeline.py:24]: What ended up happening with this?

File: frontend/components/conditions/rules.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    items: []
+  },
+  {
+    type: 'select',
+    multiple: true,
+    tag: 'Modality',
+    matchTypes: ['Any', 'All'],
+    items: ['MR', 'CT', 'PT']
+  },
+]
```
Comment (awtkns) [inline frontend/components/conditions/rules.js:16]: In the future this can be loaded from the backend

File: backend/api/routes/pipeline.py [hunk 4]
```diff
@@ -8,14 +8,34 @@
...
-from api.models.pipeline import Pipeline, PipelineLink, PipelineNode, PipelineRun, PipelineJob, PipelineJobError
+from api.models.pipeline import Pipeline, PipelineLink, PipelineNode, PipelineRun, PipelineJob, PipelineJobError, PipelineNodeCondition
 from api.schemas import pipeline as schemas
 from api.models.user import User
 from api.auth import token_auth
 
 router = APIRouter()
 
 
+# @router.get("/condition", response_model=List[schemas.PipelineCondition])
```
Comment (asim-shrestha) [inline backend/api/routes/pipeline.py:19]: Whats going on with these

File: backend/api/models/dicom.py [hunk 5]
```diff
@@ -13,6 +14,10 @@ class DicomNode(PathMixin, Base):
     input = Column(Boolean, default=False)
     output = Column(Boolean, default=False)
 
+    implementation_version_name = Column(String)
+    first_connected = Column(DateTime)
```
Comment (asim-shrestha) [inline backend/api/models/dicom.py:18]: what are these attributes for?

File: backend/api/models/pipeline.py [hunk 6]
```diff
@@ -33,6 +32,21 @@ def to_graph(self) -> DiGraph:
         return graph
 
 
+class PipelineNodeStorageBucket(PathMixin, Base):
```
Comment (asim-shrestha) [inline backend/api/models/pipeline.py:35]: Could you explain how these are supposed to work

File: backend/api/models/dicom.py (hunk 5 above)
Comment (awtkns) [reply backend/api/models/dicom.py:18]: implementation_version_name can help us 'guess' the response port. (Dicom Header)

File: backend/api/pipelining/_control.py [hunk 7]
```diff
@@ -61,91 +58,23 @@ def pipeline_run_factory(db, dicom_cls, dicom_obj_id, pipeline_id) -> PipelineRu
...
-            else:
-                print(ae_title)
-                print("ERROR: Attempted to ingest to non-existant pipeline")
-                raise NotImplementedError
+    @staticmethod
+    def ingest_globally(folder: pathlib.Path, calling_aet: str, calling_host: str, calling_port: int) -> bool:
+        return DicomIngestController._execute_task(_get_rel_folder(folder), calling_aet, calling_host, calling_port)
+
+    @staticmethod
+    def ingest_to_user(folder: pathlib.Path, calling_aet: str, calling_host: str, calling_port: int, user_id: int) -> bool:
```
Comment (asim-shrestha) [inline backend/api/pipelining/_control.py:79]: don't need one for ingest_to_pipeline?

File: backend/api/schemas/pipeline.py [hunk 8]
```diff
@@ -39,6 +49,14 @@ class PipelineNodeCreate(BaseModel):
     container_is_input: bool
     container_is_output: bool
     dicom_node_id: Optional[int]
+    conditions: Optional[List[PipelineNodeConditionCreate]] = []
+
+    @root_validator
+    def check_has_dicom_node(cls, values):
+        if values.get('container_is_input') or values.get('container_is_input'):
```
Comment (asim-shrestha) [inline backend/api/schemas/pipeline.py:56]: typo here?
```suggestion
        if values.get('container_is_input') or values.get('container_is_output'):
```

File: backend/api/schemas/pipeline.py [hunk 9]
```diff
@@ -39,6 +49,14 @@ class PipelineNodeCreate(BaseModel):
     container_is_input: bool
     container_is_output: bool
     dicom_node_id: Optional[int]
+    conditions: Optional[List[PipelineNodeConditionCreate]] = []
+
+    @root_validator
+    def check_has_dicom_node(cls, values):
```
Comment (asim-shrestha) [inline backend/api/schemas/pipeline.py:55]: don't need to change this but could maybe be more descriptive

```suggestion
    def check_io_node_has_destination(cls, values):
```

name right now suggests they should all have dicom nodes at first glance

File: backend/api/pipelining/_control.py (hunk 7 above)
Comment (asim-shrestha) [reply backend/api/pipelining/_control.py:79]: and don't both of these do the same thing

File: frontend/components/conditions/ConditionBuilder.vue [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+  },
+  data: () => ({
+    rules: rules,
+    drag: false,
+    selectedRule: undefined,
+    query: [],
+    sub: undefined,
+  }),
+  mounted() {
+    console.log(this.node)
```
Comment (asim-shrestha) [inline frontend/components/conditions/ConditionBuilder.vue:56]:
```suggestion
```

File: frontend/components/flowchart/SimpleFlowchart.vue [hunk 11]
```diff
@@ -150,15 +169,12 @@ export default {
...
   methods: {
     setDestinations(destination) {
-      const index = this.pipelineNodeDestinations.findIndex(
-        pipelineDestination =>
-          pipelineDestination.pipelineNodeId === destination.pipelineNodeId
-      )
-      if (index >= 0) {
-        this.pipelineNodeDestinations.splice(index, 1)
-      }
-      this.pipelineNodeDestinations.push(destination)
```
Comment (asim-shrestha) [inline frontend/components/flowchart/SimpleFlowchart.vue:160]: was this incorrect before? looks like it always pushed the destination

File: backend/api/dicom/scp.py [hunk 12]
```diff
@@ -117,8 +125,8 @@ def __init__(self, ae_title='PICOM_SCP', host='localhost', port=11112, debug=Fal
         self._ae.supported_contexts = AllStoragePresentationContexts
         self._ae.add_supported_context(VerificationSOPClass)
 
-        if debug:
-            debug_logger()
+        # if debug:
```
Comment (asim-shrestha) [inline backend/api/dicom/scp.py:128]: should we uncomment

**Pull request merged**